File: Retail_Pulse/src/services/inventoryService.ts
import api from "./api";

// ------------------------------------
// Get Inventory
// ------------------------------------

export const getInventory = async (params?: {
  search?: string;
  category?: string;
  brand?: string;
  stock_status?: string;
}) => {
  const query: Record<string, string> = {};

  if (params?.search?.trim()) {
    query.search = params.search;
  }

  if (params?.category?.trim()) {
    query.category = params.category;
  }

  if (params?.brand?.trim()) {
    query.brand = params.brand;
  }

  if (params?.stock_status?.trim()) {
    query.stock_status = params.stock_status;
  }

  const response = await api.get("/inventory/", {
    params: query,
  });

  return response.data;
};

// ------------------------------------
// Dashboard Summary
// ------------------------------------

export const getInventorySummary = async () => {
  const response = await api.get(
    "/inventory/dashboard/summary"
  );

  return response.data;
};

// ------------------------------------
// Movement History
// ------------------------------------

export const getMovementHistory = async (
  inventoryId: number
) => {
  const response = await api.get(
    `/inventory/${inventoryId}/history`
  );

  return response.data;
};

// ------------------------------------
// Add Stock
// ------------------------------------

export const addStock = async (
  inventoryId: number,
  data: {
    quantity: number;
    reason: string;
    remarks?: string;
  }
) => {
  const response = await api.put(
    `/inventory/${inventoryId}/add-stock`,
    {
      movementType: "add",
      quantity: data.quantity,
      reason: data.reason,
      remarks: data.remarks,
    }
  );

  return response.data;
};

// ------------------------------------
// Remove Stock
// ------------------------------------

export const removeStock = async (
  inventoryId: number,
  data: {
    quantity: number;
    reason: string;
    remarks?: string;
  }
) => {
  const response = await api.put(
    `/inventory/${inventoryId}/remove-stock`,
    {
      movementType: "remove",
      quantity: data.quantity,
      reason: data.reason,
      remarks: data.remarks,
    }
  );

  return response.data;
};

// ------------------------------------
// Manual Adjustment
// ------------------------------------

export const adjustStock = async (
  inventoryId: number,
  data: {
    quantity: number;
    reason: string;
    remarks?: string;
  }
) => {
  const response = await api.put(
    `/inventory/${inventoryId}/adjust`,
    {
      movementType: "adjust",
      quantity: data.quantity,
      reason: data.reason,
      remarks: data.remarks,
    }
  );

  return response.data;
};

// ------------------------------------
// Create Inventory
// ------------------------------------

export const createInventory = async (data: {
  productId: number;
  currentStock: number;
  reservedStock: number;
  reorderLevel: number;
}) => {
  const response = await api.post(
    "/inventory/",
    data
  );

  return response.data;
};

File: backend/app/services/inventory_service.py
from sqlalchemy.orm import Session
from sqlalchemy import or_

from app.models.inventory import Inventory
from app.models.inventory_movement import InventoryMovement
from app.models.product import Product

from app.schemas.inventory_schema import InventoryCreate


class InventoryService:

    @staticmethod
    def calculate_status(available_stock: int, reorder_level: int):

        if available_stock == 0:
            return "Out of Stock"

        if available_stock <= reorder_level:
            return "Low Stock"

        return "In Stock"

    @staticmethod
    def create_inventory(
        db: Session,
        company_id: int,
        data: InventoryCreate,
    ):

        available = data.currentStock - data.reservedStock

        status = InventoryService.calculate_status(
            available,
            data.reorderLevel,
        )

        inventory = Inventory(
            companyId=company_id,
            productId=data.productId,
            currentStock=data.currentStock,
            reservedStock=data.reservedStock,
            availableStock=available,
            reorderLevel=data.reorderLevel,
            stockStatus=status,
        )

        db.add(inventory)
        db.commit()
        db.refresh(inventory)

        return inventory

    @staticmethod
    def get_all_inventory(
        db: Session,
        company_id: int,
        search: str = None,
        category: int = None,
        brand: str = None,
        stock_status: str = None,
    ):

        query = (
    db.query(Inventory, Product)
    .join(
        Product,
        Inventory.productId == Product.id,
    )
    .join(Product.category)
    .filter(
        Inventory.companyId == company_id
    )
)

        if search:
            query = query.filter(
                or_(
                    Product.name.ilike(f"%{search}%"),
                    Product.sku.ilike(f"%{search}%"),
                )
            )

        if category:
            query = query.filter(
                Product.categoryId == category
            )

        if brand:
            query = query.filter(
                Product.brand == brand
            )

        if stock_status:
            query = query.filter(
                Inventory.stockStatus == stock_status
            )

        result = []

        for inventory, product in query.all():

            result.append(
                {
                    "id": inventory.id,
                    "companyId": inventory.companyId,
                    "productId": inventory.productId,

                    "productName": product.name,
                    "sku": product.sku,
                    "category": (product.category.name if product.category else ""),
                    "brand": product.brand,

                    "currentStock": inventory.currentStock,
                    "reservedStock": inventory.reservedStock,
                    "availableStock": inventory.availableStock,
                    "reorderLevel": inventory.reorderLevel,
                    "stockStatus": inventory.stockStatus,
                    "updatedAt": inventory.updatedAt,
                }
            )

        return result

    @staticmethod
    def add_stock(
        db: Session,
        inventory: Inventory,
        quantity: int,
        reason: str,
        remarks: str,
        user_id: int,
    ):

        previous = inventory.currentStock

        inventory.currentStock += quantity

        inventory.availableStock = (
            inventory.currentStock
            - inventory.reservedStock
        )

        inventory.stockStatus = InventoryService.calculate_status(
            inventory.availableStock,
            inventory.reorderLevel,
        )

        movement = InventoryMovement(
            inventoryId=inventory.id,
            movementType="Stock Addition",
            quantityChanged=quantity,
            previousQuantity=previous,
            updatedQuantity=inventory.currentStock,
            reason=reason,
            remarks=remarks,
            performedBy=user_id,
        )

        db.add(movement)
        db.commit()
        db.refresh(inventory)

        return inventory

    @staticmethod
    def remove_stock(
        db: Session,
        inventory: Inventory,
        quantity: int,
        reason: str,
        remarks: str,
        user_id: int,
    ):

        if quantity > inventory.availableStock:
            raise Exception("Insufficient stock.")

        previous = inventory.currentStock

        inventory.currentStock -= quantity

        inventory.availableStock = (
            inventory.currentStock
            - inventory.reservedStock
        )

        inventory.stockStatus = InventoryService.calculate_status(
            inventory.availableStock,
            inventory.reorderLevel,
        )

        movement = InventoryMovement(
            inventoryId=inventory.id,
            movementType="Stock Removal",
            quantityChanged=quantity,
            previousQuantity=previous,
            updatedQuantity=inventory.currentStock,
            reason=reason,
            remarks=remarks,
            performedBy=user_id,
        )

        db.add(movement)
        db.commit()
        db.refresh(inventory)

        return inventory

    @staticmethod
    def manual_adjustment(
        db: Session,
        inventory: Inventory,
        quantity: int,
        reason: str,
        remarks: str,
        user_id: int,
    ):

        previous = inventory.currentStock

        inventory.currentStock = quantity

        inventory.availableStock = (
            inventory.currentStock
            - inventory.reservedStock
        )

        inventory.stockStatus = InventoryService.calculate_status(
            inventory.availableStock,
            inventory.reorderLevel,
        )

        movement = InventoryMovement(
            inventoryId=inventory.id,
            movementType="Manual Adjustment",
            quantityChanged = quantity - previous,
            previousQuantity=previous,
            updatedQuantity=inventory.currentStock,
            reason=reason,
            remarks=remarks,
            performedBy=user_id,
        )

        db.add(movement)
        db.commit()
        db.refresh(inventory)

        return inventory

    @staticmethod
    def movement_history(
        db: Session,
        inventory_id: int,
    ):

        return (
            db.query(InventoryMovement)
            .filter(
                InventoryMovement.inventoryId == inventory_id
            )
            .order_by(
                InventoryMovement.createdAt.desc()
            )
            .all()
        )

    @staticmethod
    def dashboard_summary(
        db: Session,
        company_id: int,
    ):

        inventory = (
            db.query(Inventory)
            .filter(
                Inventory.companyId == company_id
            )
            .all()
        )

        total_products = len(inventory)

        total_inventory = sum(
            i.currentStock
            for i in inventory
        )

        low_stock = len(
            [
                i
                for i in inventory
                if i.stockStatus == "Low Stock"
            ]
        )

        out_stock = len(
            [
                i
                for i in inventory
                if i.stockStatus == "Out of Stock"
            ]
        )

        return {
            "totalProducts": total_products,
            "totalInventory": total_inventory,
            "lowStockProducts": low_stock,
            "outOfStockProducts": out_stock,
        }